internal/stacks/stackruntime/internal/stackeval/component_config.go

```diff
@@ -330,19 +314,55 @@ func (c *ComponentConfig) CheckProviders(ctx context.Context, phase EvalPhase) (
 			continue
 		}
 
+		// Next, we want to make sure the linked providers are actually of the
+		// same type.
+
 		const errSummary = "Invalid provider configuration"
 		if actualTy := result.Value.Type(); stackconfigtypes.IsProviderConfigType(actualTy) {
 			// Then we at least got a provider reference of some kind.
 			actualTypeAddr := stackconfigtypes.ProviderForProviderConfigType(actualTy)
 			if actualTypeAddr != typeAddr {
+				var errorDetail string
+
+				stackName, matchingTypeExists := stackConfig.ProviderLocalName(ctx, typeAddr)
+				_, matchingNameExists := stackConfig.ProviderForLocalName(ctx, componentAddr.LocalName)
+				moduleProviderTypeExplicit := elem.Value.Explicit
+				if !matchingTypeExists && !matchingNameExists {
+					// Then the user just hasn't declared the target provider
+					// type or name at all. We'll return a generic error message
+					// asking the user to update the required_providers list.
+					errorDetail = "\n\nDeclare the required provider in the stack's required_providers block, and then assign a configuration for that provider in this component's \"providers\" argument."
+				} else if !matchingNameExists {
+					// Then we have a type that matches, but the name doesn't.
+					errorDetail = fmt.Sprintf("\n\nThis stack has a configured provider of the right type under the name %q. Update this component's \"providers\" argument to reference this provider.", stackName)
```

Comment:
Very tiny nit: we tend to use "correct" rather than "right" because "right" has other meanings and so is potentially harder to understand for the reader.